File: Level 2/롤케이크 자르기/solution.py
from collections import Counter


def solution(toppings):
    answer = 0

    right_topping_counter = Counter(toppings)
    left_topping_set = set()

    for topping in toppings:
        left_topping_set.add(topping)
        right_topping_counter[topping] -= 1

        if right_topping_counter[topping] == 0:
            right_topping_counter.pop(topping)

        if len(left_topping_set) == len(right_topping_counter):
            answer += 1

    return answer

File: Level 2/롤케이크 자르기/solution.js
function solution(toppings) {
  let answer = 0;

  const rightToppingMap = new Map();
  const leftToppingSet = new Set();

  toppings.forEach((topping) => {
    rightToppingMap.set(topping, (rightToppingMap.get(topping) || 0) + 1);
  });

  toppings.forEach((topping) => {
    leftToppingSet.add(topping);
    rightToppingMap.set(topping, rightToppingMap.get(topping) - 1);

    if (rightToppingMap.get(topping) === 0) {
      rightToppingMap.delete(topping);
    }

    if (leftToppingSet.size === rightToppingMap.size) {
      answer++;
    }
  });

  return answer;
}
